sql/core/src/main/scala/org/apache/spark/sql/sources/interfaces.scala

```diff
@@ -683,9 +685,29 @@ abstract class HadoopFsRelation private[sql](
       }
     }
 
-    buildInternalScan(requiredColumns, filters, inputStatuses, broadcastedConf)
+    if (bucketSpec.isDefined && sqlContext.conf.bucketingEnabled()) {
+      val inputFilesWithBucketId = allInputFiles.groupBy(getBucketId)
```

Comment:
or we can combine the validation and `groupBy` stuff? i.e. return a `Option` here, if it's `None`, we just log error and continue without bucketing, if it `Some`, go to the bucketing branch.